File: my-solutions/0347-top-k-frequent-elements/solution.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        res = []
        num_ctr = Counter(nums)
        rev_map = defaultdict(list)
        for key, val in num_ctr.items():
            rev_map[val].append(key)

        freqs = sorted(list(rev_map.keys()), reverse = True)
        
        for freq in freqs:
            for val in rev_map[freq]:
                res.append(val)
                k -= 1
                if k <= 0:
                    return res
        return res

        

File: my-solutions/0347-top-k-frequent-elements/solution.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> num_freq;
        for(auto& num: nums) {
            ++num_freq[num];
        }
        map<int, vector<int>, greater<int>> freq_num;
        for(auto& p: num_freq) {
            freq_num[p.second].push_back(p.first);
        }

        vector<int> res;
        for(auto& p: freq_num) {
            for(auto& num: p.second) {
                res.push_back(num);
                if(res.size() == k) {
                    return res;
                }
            }
        }
        return res;
    }
};
